Docs 0cc bench (#15)
wphicks/cuml: docs/source/zero-code-change-benchmarks.rst

cuml.accel: Zero Code Change Acceleration Benchmarks
====================================================

Training
--------

> One of the biggest advantages of cuML is how much faster training becomes, which is true as well when using the cuML Accelerator, with no code changes!

Reviewer: I'd be interested to get feedback from others about the overall tone here. In general for docs, I think I'd favor a somewhat more factual focus rather than opening with a pitch for the overall project. It can still be enthusiastic and highlight our successes without veering toward marketing. Maybe we could instead say something along the lines of "cuML offers 2-80 times faster training for most algorithms, especially on large datasets. With cuml.accel, you can enable the same sort of acceleration on existing Scikit-Learn, UMAP, and HDBScan scripts without changing a line of code." Then the following paragraph could be reframed as adding detail to this assertion by saying what we would expect for each individual algorithm.

Author: I think you are fully correct, a phrasing that is more doc oriented, I kind of was writing blog style

Reviewer 2: I think the tone should still be familiar and conversational, but I agree with @wphicks , a slightly more "objective" / "factual" tone is probably preferred here.

> One of the biggest advantages of cuML is how much faster training becomes, which is true as well when using the cuML Accelerator, with no code changes!

Reviewer 3: ```suggestion
One of the biggest advantages of cuML is how much faster training becomes. This is true as well when using the cuML Accelerator, with no code changes!
```

> Algorithms like HDBSCAN, t-SNE, and UMAP often see speedups anywhere from 60× to over 300×, transforming tasks that could take hours or even days into minutes. Even more conventional methods such as KMeans and Random Forests can achieve training speedups of 15× to 80×, while models like LogisticRegression, Lasso, PCA, and Ridge still benefit significantly (typically in the 2×–10× range). The following benchmarks are a summary of the type of speedups that can be seen in datasets that range from 8 to 512 columns.

Reviewer: ```suggestion

cuML offers accelerated inference and training for classical ML models using NVIDIA GPUs. With cuml.accel, you can get the benefit of similar acceleration in existing Scikit-Learn, UMAP, and HDBScan scripts without changing a line of code. While the exact speedup depends on the model, dataset size, and hyperparameters, the following benchmarks should give a general sense of the benefit you're likely to observe when using `cuml.accel.`

Training
--------

While both training and inference benefit from GPU acceleration, cuML tends to offer even more significant gains for training than inference. Training times 2-80 times faster are typical for cuML, especially with large datasets. `cuml.accel` offers similar speedups without requiring any cuML-specific code changes.

Relatively complex manifold algorithms like HDBSCAN, t-SNE, and UMAP tend to see the largest benefit from `cuml.accel`. Speedups of 60x to 300x are typical for realistic workloads. Simpler algorithms like KMeans and Random Forest can also see speedups of 15x to 80x. Even the simplest ML algorithms, like Logistic Regression, Lasso, PCA, and Ridge will typically have speedusp of 2x to 10x.

In the following chart, we see the relative speedup obtained by running the same code with and without `cuml.accel`. The datasets for these workloads range in width from 8 to 512 features. As we can see `cuml.accel` offered the most benefit for HDBSCAN, with a 179x speedup, but even KNeighborsRegressor saw a 2x speedup.
```

.. image:: img/overall_speedup.png
   :alt: Overall speedup

Comparison of cuML and cuml.accel
---------------------------------

> One question about cuml.accel is: how does it compare to using cuML directly? The answer varies, but in cases where the models are compared with the same parameters and data sizes, the overhead is quite small when it comes to training models:

Reviewer: ```suggestion
While cuml.accel tries to provide as much acceleration as cuML-specific scripts, there is some overhead relative to direct cuML invocations. While the exact amount of overhead depends on the estimator, parameters, and data size, the overhead is typically quite low for model training:
```

> One question about cuml.accel is: how does it compare to using cuML directly? The answer varies, but in cases where the models are compared with the same parameters and data sizes, the overhead is quite small when it comes to training models:

Reviewer 3: Should we format `cuml.accel` as code instead of prose?

```suggestion
One question about ``cuml.accel`` is: how does it compare to using cuML directly? The answer varies, but in cases where the models are compared with the same parameters and data sizes, the overhead is quite small when it comes to training models:
```

.. image:: img/overall_overhead.png
   :alt: Overall overhead

> Two important aspects arise immediately: training typically is quite computationally expensive, so the cost of transferring data from CPU and GPU and the machinery of cuml.accel have overheads that don't affect the acceleration and running time significantly. But even here, one can immediately notice that the overhead is more significant for tasks that are simpler, for example training KNeighbors modeks. There, using cuML directly can be significantly faster if one wants to get the maximum performance of a GPU, though it's important to note that the difference in execution time is the difference of computing in seconds vs miliseconds.

Reviewer: I'd recommend rewriting this paragraph for simplicity and clarity. It's not obvious to me until partway through the paragraph what it's trying to tell me. Maybe we put the bottom line up front: "Simpler estimators tend to have relatively more overhead with cuml.accel since they already run quite fast." And then we can go on to provide detail about that.

> Two important aspects arise immediately: training typically is quite computationally expensive, so the cost of transferring data from CPU and GPU and the machinery of cuml.accel have overheads that don't affect the acceleration and running time significantly. But even here, one can immediately notice that the overhead is more significant for tasks that are simpler, for example training KNeighbors modeks. There, using cuML directly can be significantly faster if one wants to get the maximum performance of a GPU, though it's important to note that the difference in execution time is the difference of computing in seconds vs miliseconds.

Reviewer 3: ```suggestion
Two important aspects arise immediately: training typically is quite computationally expensive, so the cost of transferring data from CPU to GPU and the machinery of ``cuml.accel`` have overheads that don't affect the acceleration and running time significantly. But even here, one can immediately notice that the overhead is more significant for tasks that are simpler, for example training ``KNeighbors`` models. There, using cuML directly can be significantly faster if one wants to get the maximum performance of a GPU, though it's important to note that the difference in execution time is the difference of computing in seconds vs milliseconds.
```

> It’s also important to note how dataset shape influences these gains. For skinny datasets—where you have relatively few features but many rows—GPU acceleration still provides a great performance boost, although the relative advantage may be more modest for simpler algorithms that are already quite fast on CPU.

Reviewer 3: ```suggestion
It’s also important to note how dataset shape influences these gains. For skinny datasets — where you have relatively few features but many rows — GPU acceleration still provides a great performance boost, although the relative advantage may be more modest for simpler algorithms that are already quite fast on CPU.
```

.. image:: img/skinny_speedup.png
   :alt: Skinny speedup

> Wide datasets, on the other hand, truly showcase the accelerator’s strengths. High-dimensional tasks often require intense computation and can bog down CPU-based workflows. In these cases, the cuML Accelerator steps in to deliver some of its most dramatic speedups, especially for dimension reduction methods (t-SNE, UMAP) and other math-heavy operations. It's not uncommon that a task that was unfeasible to achieve before, like incorporating UMAP and HDBSCAN in complex, high dimensional workflows, can now easily be achieved thanks to cuML and cuml.accel:

Reviewer 3: ```suggestion
Wide datasets, on the other hand truly showcase the accelerator’s strengths. High-dimensional tasks often require intense computation and can bog down CPU-based workflows. In these cases, the cuML Accelerator steps in to deliver some of its most dramatic speedups, especially for dimension reduction methods (t-SNE, UMAP) and other math-heavy operations. It's not uncommon that a task that was unfeasible to achieve before, like incorporating UMAP and HDBSCAN in complex, high dimensional workflows, can now easily be achieved thanks to cuML and ``cuml.accel``:
```

.. image:: img/wide_speedup.png
   :alt: Wide speedup

Inference
----------

> While the accelerator also speeds up inference, the gains tend to be smaller in absolute terms because inference is usually much faster than training to begin with. Still, seeing a 2×–7× improvement (as with KNeighbors or RandomForest) can be critical if you’re running large-scale or real-time predictions.  However, for large-batch or repeated inference scenarios, the GPU acceleration can provide significant value.

Reviewer: This is really good. I like the clear statement at the top with detail following. Let's follow the same pattern for the training overhead discussion above.

.. image:: img/inference_speedup.png
   :alt: Inference Speedup

> The main caveat is that data transfer overhead becomes a bigger slice of total runtime when inference itself is quick. If you’re doing tiny batches of predictions one at a time, the overhead can reduce your net benefit, and the speed between CPU and GPU can be withing statistical margin of error. The overhead can be significant in that case, in which using cuML directly with GPU inputs (like CuPy) can provide a significant advantage over using cuml.accel.

Reviewer 4: ```suggestion
The main caveat is that data transfer overhead becomes a bigger slice of total runtime when inference itself is quick. If you’re doing tiny batches of predictions one at a time, the overhead can reduce your net benefit, and the speed between CPU and GPU can be within statistical margin of error. The overhead can be significant in that case, in which using cuML directly with GPU inputs (like CuPy) can provide a significant advantage over using cuml.accel.
```

In the following chart, we see the relative speedup obtained by running the same code with and without `cuml.accel`. The datasets for these workloads range in width from 8 to 512 features. As we can see `cuml.accel` offered the most benefit for HDBSCAN, with a 179x speedup, but even KNeighborsRegressor saw a 2x speedup.

.. image:: img/overall_speedup.png
   :alt: Overall speedup

What’s the overhead compared to invoking cuML directly?
-------------------------------------------------------

While cuml.accel tries to provide as much acceleration as cuML-specific scripts, there is some overhead relative to direct cuML invocations. So one might reasonably wonder at what point it makes sense to rewrite code and invoke cuML directly to squeeze out every bit of performance. While the exact amount of overhead depends on the estimator, parameters, and data size, the overhead is typically quite low for model training, some algorithms have a bit more overhead than others:

.. image:: img/overall_overhead.png
   :alt: Overall overhead

> The differences can be attributed to one main factr: training typically is quite computationally expensive. So the cost of transferring data from CPU to GPU and the machinery of the cuML Accelerator overheads don't affect the runtime significantly. But even here, one can immediately notice that the overhead is more significant for tasks that are simpler, for example training ``KNeighbors`` models. There, using cuML directly can be significantly faster if one wants to get the maximum performance of a GPU, though it's important to note that the difference in execution time is the difference of computing in seconds vs milliseconds.

Reviewer 2: One typo, one grammar error, everything else are suggestions:

```suggestion
The differences can be attributed to one main factor: training is typically computationally expensive. So the cost of transferring data from CPU to GPU and any minor overheads incurred by cuml.accel doesn't have any significant effect on the overall runtime. But even here, one can immediately notice that the overhead is more significant for tasks that are simpler, for example training ``KNeighbors`` models. There, invoking cuML explicitly can be significantly faster if one wants to get the maximum performance of a GPU. Though it's important to note that the difference in execution time is the difference of computing in seconds vs milliseconds.
```

It’s also important to note how dataset shape influences these gains. For skinny datasets — where you have relatively few features but many rows — GPU acceleration still provides a great performance boost, although the relative advantage may be more modest for simpler algorithms that are already quite fast on CPU. The following benchmark shows speedups for datasets with 8 and 16 features:

.. image:: img/skinny_speedup.png
   :alt: Skinny speedup

Wide datasets, on the other hand truly showcase the accelerator’s strengths. High-dimensional tasks often require intense computation and can bog down CPU-based workflows. In these cases, the cuML Accelerator steps in to deliver some of its most dramatic speedups, especially for dimension reduction methods (t-SNE, UMAP) and other math-heavy operations. It's not uncommon that a task that was unfeasible to achieve before, like incorporating UMAP and HDBSCAN in complex, high dimensional workflows, can now easily be achieved thanks to cuML and ``cuml.accel``. The following benchmark shows those speedups for datasets with 128, 256 and 512 features:

.. image:: img/wide_speedup.png
   :alt: Wide speedup

Inference
----------

> While the accelerator also speeds up inference, the gains tend to be smaller in absolute terms because inference is usually much faster than training to begin with. Still, a 2×–7× improvement (as with KNeighbors or RandomForest) can be critical ifor running large-scale or real-time predictions.  Especially for large-batch or repeated inference scenarios, the GPU acceleration can provide significant value.

Reviewer 2: ```suggestion
While the accelerator also speeds up inference, the gains tend to be smaller in absolute terms because inference is usually much faster than training to begin with. Still, a 2×–7× improvement (as with KNeighbors or RandomForest) can be critical for running large-scale or real-time predictions.  Especially for large-batch or repeated inference scenarios, the GPU acceleration can provide significant value.
```
